### PR TITLE
compactor: Add offline deduplication

### LINKED ISSUE #1014: Compact: Offline deduplication
Currently Thanos provides dedup function in Query component, it can merge the data on the fly from Prometheus HA pairs or remote object storage.
However, the Query dedup function is not query efficiently, because the replica's blocks in tsdb/object storage are not really merged so that the Query API has to load duplicated data in each request.

With current Prometheus TSDB design, it seems difficult to implement dedup function to merge blocks in each TSDB node on Thanos side.
However, it should be easily to implement it for the object storage as it is one centralized storage, and also we already have one Compactor component running on it.

Offering dedup function on object storage side will definitely help fast metrics query latency and reduce the object storage cost.

Do we have any plan to support such requirements in Thanos? If had, I would like to know your ideas about this feature.

Thanks.

### PR DESCRIPTION
**Pull request by yeya24**
Signed-off-by: yeya24 <yb532204897@gmail.com>

Fixes #1014 

* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

Another attempt on offline deduplication using the existing querier's penalty deduplication algorithm.

~~This pr is WIP as I need to do more testing and is currently blocked by https://github.com/prometheus/prometheus/pull/8836, so I use my own fork here.~~

But as mentioned in https://github.com/thanos-io/thanos/issues/1014#issuecomment-746062533, this deduplication algorithm works for 99% of cases but is wrong in 1% case. So it has its own risk.

## Verification

Tested with 2 Prometheus replicas running in k8s. I replicate the minio bucket to a local directory as a filesystem objstore for testing. The local directory is used for running dedup compactor. There are 12h test data and I run 1 thanos store and 1 thanos querier for each bucket.

Undeduped bucket:

Query deduplication has to be enabled.

Screenshot from 2021-05-17 11-09-16

Bucket size: 41m (36m if normal compaction is enabled)

Offline deduped bucket:

Query deduplication disabled.

Screenshot from 2021-05-17 11-10-36

Bucket size: 19m

### CODE REVIEW

**Comment by yeya24** [comment]
I think the failed E2E test is related to my change, but I am not sure why.

File: cmd/thanos/compact.go [hunk 1]
```diff
@@ -623,9 +638,13 @@ func (cc *compactConfig) registerFlag(cmd extkingpin.FlagClause) {
 
 	cmd.Flag("compact.enable-vertical-compaction", "Experimental. When set to true, compactor will allow overlaps and perform **irreversible** vertical compaction. See https://thanos.io/tip/components/compact.md/#vertical-compactions to read more."+
 		"Please note that this uses a NAIVE algorithm for merging (no smart replica deduplication, just chaining samples together)."+
-		"NOTE: This flag is ignored and (enabled) when --deduplication.replica-label flag is set.").
+		"NOTE: This flag is ignored and (enabled) when --deduplication.replica-label or --compact.enable-dedup flag is set.").
```
Comment (bwplotka) [inline cmd/thanos/compact.go:641]: Why we need 3rd flag for enabling dedup? (:

File: cmd/thanos/compact.go [hunk 2]
```diff
@@ -623,9 +638,13 @@ func (cc *compactConfig) registerFlag(cmd extkingpin.FlagClause) {
...
 		Hidden().Default("false").BoolVar(&cc.enableVerticalCompaction)
 
+	cmd.Flag("compact.enable-dedup", "Experimental. When set to true, compactor will use the penalty algorithm to merge chunk series. At least one --deduplication.replica-label has to be set in this mode.").
+		Default("false").BoolVar(&cc.enableDedup)
+
+	// Update this. This flag works for both dedup version compactor and the original compactor.
 	cmd.Flag("deduplication.replica-label", "Label to treat as a replica indicator of blocks that can be deduplicated (repeated flag). This will merge multiple replica blocks into one. This process is irreversible."+
```
Comment (bwplotka) [inline cmd/thanos/compact.go:648]: Should we unhide this at some point? 🤔

File: cmd/thanos/compact.go [hunk 3]
```diff
@@ -623,9 +638,13 @@ func (cc *compactConfig) registerFlag(cmd extkingpin.FlagClause) {
 
 	cmd.Flag("compact.enable-vertical-compaction", "Experimental. When set to true, compactor will allow overlaps and perform **irreversible** vertical compaction. See https://thanos.io/tip/components/compact.md/#vertical-compactions to read more."+
 		"Please note that this uses a NAIVE algorithm for merging (no smart replica deduplication, just chaining samples together)."+
-		"NOTE: This flag is ignored and (enabled) when --deduplication.replica-label flag is set.").
+		"NOTE: This flag is ignored and (enabled) when --deduplication.replica-label or --compact.enable-dedup flag is set.").
 		Hidden().Default("false").BoolVar(&cc.enableVerticalCompaction)
 
+	cmd.Flag("compact.enable-dedup", "Experimental. When set to true, compactor will use the penalty algorithm to merge chunk series. At least one --deduplication.replica-label has to be set in this mode.").
```
Comment (bwplotka) [inline cmd/thanos/compact.go:644]: Bool flags are not very flexible, you see that it makes us add tons of them. Imagine we have 3rd merge algorithm. 

What about having an enum flag for all algorithms? This will allow us to extend it easier in future (:

File: pkg/compact/dedup.go [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+	"container/heap"
+	"math"
+
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/tsdb/chunkenc"
+	"github.com/prometheus/prometheus/tsdb/chunks"
+)
+
+// NewDedupChunkSeriesMerger merges several chunk series into one.
```
Comment (bwplotka) [inline pkg/compact/dedup.go:16]: I assume you copied some of things over to next file? Can you mark which of those functions/methods are unchanged?

File: cmd/thanos/compact.go [hunk 5]
```diff
@@ -301,9 +303,21 @@ func runCompact(
...
+		enableVerticalCompaction = true
+
+		if len(conf.dedupReplicaLabels) == 0 {
+			return errors.New("dedup mode needs at least one replica label specified")
+		}
+	}
 	// Instantiate the compactor with different time slices. Timestamps in TSDB
 	// are in milliseconds.
-	comp, err := tsdb.NewLeveledCompactor(ctx, reg, logger, levels, downsample.NewPool())
+	comp, err := tsdb.NewLeveledCompactor(ctx, reg, logger, levels, downsample.NewPool(), mergeFunc)
```
Comment (bwplotka) [inline cmd/thanos/compact.go:320]: You are one step ahead! Nice 🤗

File: pkg/compact/dedup.go (hunk 4 above)
Comment (bwplotka) [reply pkg/compact/dedup.go:16]: I assume nothing has changed here

**Review by bwplotka — DISMISSED** [review]
This is epic, If not for the file move this would have been 50 line change to add something so powerful. Well done!

I have a strong suggestion regarding the bool flag, otherwise LGTM!

One extra question: Are we sure penalty algo will work properly on 1:1 dups? (: (extra test would help)

File: pkg/compact/dedup.go (hunk 4 above)
Comment (yeya24) [reply pkg/compact/dedup.go:16]: This is mainly a cleanup and I am also wondering is there a better way to do it.
1. The original files only works for SeriesSet and the compact merger function requires ChunkSeriesSet. So I cannot directly use the original functions.
2. The original functions have logic for counter reset and adjustment. This is not needed in the compactor side.
3. I need a place to put some code for the encoder and decoder of chunk and series. Do you think it reasonable if I put them in the `query` package?

File: cmd/thanos/compact.go (hunk 3 above)
Comment (yeya24) [reply cmd/thanos/compact.go:644]: Yes, that STGM

File: cmd/thanos/compact.go (hunk 1 above)
Comment (yeya24) [reply cmd/thanos/compact.go:641]: If we add a dedup algorithm enum flag then this can be removed for sure.

File: cmd/thanos/compact.go (hunk 2 above)
Comment (yeya24) [reply cmd/thanos/compact.go:648]: Agree

File: pkg/compact/dedup.go (hunk 4 above)
Comment (bwplotka) [reply pkg/compact/dedup.go:16]: Well we already have some code for there, but you are right, it's something both compaction and querier would use. Maybe shared dedup package would do?

Comment (yeya24) [reply pkg/compact/dedup.go:16]: Yes, a shared package makes sense.

Comment (yeya24) [reply pkg/compact/dedup.go]: As this change is mainly a cleanup, I split it to a new pr https://github.com/thanos-io/thanos/pull/4242.

**Comment by yeya24** [comment]
This works for downsampled blocks now. Each qurier runs against one bucket. The two buckets contain the same downsample level1 blocks and one bucket is deduplicated then.


Screenshot from 2021-05-17 20-40-38

**Review by kakkoyun — COMMENTED** [review]
Looking good. I've just skimmed, I need to spend some more time reviewing this.

File: pkg/dedup/chunk_iter.go [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/tsdb/chunkenc"
+	"github.com/prometheus/prometheus/tsdb/chunks"
+
+	"github.com/thanos-io/thanos/pkg/compact/downsample"
+)
+
+// NewDedupChunkSeriesMerger merges several chunk series into one.
+// Deduplication is based on penalty based deduplication algorithm without handling counter reset.
+func NewDedupChunkSeriesMerger() storage.VerticalChunkSeriesMergeFunc {
```
[the remark this answers is not in the document]
Comment (yeya24) [reply pkg/dedup/chunk_iter.go:19]: That's a different function for chunk series. Yes I would update the name. Thanks!

File: cmd/thanos/compact.go [hunk 7]
```diff
@@ -302,9 +305,20 @@ func runCompact(
...
 	}()
+
+	var mergeFunc storage.VerticalChunkSeriesMergeFunc
+	switch conf.dedupFunc {
+	case compact.DedupAlgorithmPenalty:
+		mergeFunc = dedup.NewChunkSeriesMerger()
+
+		if len(conf.dedupReplicaLabels) == 0 {
+			return errors.New("penalty based deduplication needs at least one replica label specified")
+		}
```
Comment (bwplotka) [inline cmd/thanos/compact.go:316]: Can we error out on default, so if someone puts something not supported?

File: cmd/thanos/compact.go [hunk 8]
```diff
@@ -623,14 +638,19 @@ func (cc *compactConfig) registerFlag(cmd extkingpin.FlagClause) {
 		Default("48h").SetValue(&cc.deleteDelay)
 
 	cmd.Flag("compact.enable-vertical-compaction", "Experimental. When set to true, compactor will allow overlaps and perform **irreversible** vertical compaction. See https://thanos.io/tip/components/compact.md/#vertical-compactions to read more."+
-		"Please note that this uses a NAIVE algorithm for merging (no smart replica deduplication, just chaining samples together)."+
+		"Please note that by default this uses a NAIVE algorithm for merging. If you need a smarter deduplication algorithm, please set it via -- compact.dedup-func."+
```
Comment (bwplotka) [inline cmd/thanos/compact.go:641]:
```suggestion
		"Please note that by default this uses a NAIVE algorithm for merging. If you need a different deduplication algorithm (e.g one that works well with Prometheus replicas), please set it via -- compact.dedup-func."+
```

Comment (bwplotka) [reply cmd/thanos/compact.go:641]: Can we refer to docs: https://thanos.io/tip/components/compact.md/#vertical-compaction-use-cases

File: cmd/thanos/compact.go [hunk 9]
```diff
@@ -623,14 +638,19 @@ func (cc *compactConfig) registerFlag(cmd extkingpin.FlagClause) {
...
 		"NOTE: This flag is ignored and (enabled) when --deduplication.replica-label flag is set.").
 		Hidden().Default("false").BoolVar(&cc.enableVerticalCompaction)
 
+	cmd.Flag("compact.dedup-func", "Experimental. Deduplication algorithm for merging overlapping blocks. "+
+		"Possible values are: \"\", \"penalty\". If no value is specified, the default compact deduplication merger is used, which performs 1:1 deduplication for samples. "+
+		"When set to penalty, penalty based deduplication algorithm will be used. At least one replica label has to be set via --deduplication.replica-label flag.").
```
Comment (bwplotka) [inline cmd/thanos/compact.go:647]: Can we refer to docs? https://thanos.io/tip/components/compact.md/#vertical-compaction-use-cases

File: pkg/compact/compact.go [hunk 10]
```diff
@@ -41,6 +41,12 @@ const (
 	ResolutionLevel1h  = ResolutionLevel(downsample.ResLevel2)
 )
 
+const (
+	// DedupAlgorithmPenalty is the penalty based compactor series merge algorithm.
+	// This is the same as the online deduplication of querier except counter reset handling.
+	DedupAlgorithmPenalty = "penalty"
```
Comment (bwplotka) [inline pkg/compact/compact.go:47]: Can we update https://thanos.io/tip/components/compact.md/#vertical-compaction-use-cases to name this algorithm `penalty`? We used to call it `realistic` (I think penalty based name is nice as it's very specific)

File: pkg/dedup/chunk_iter.go [hunk 11]
```diff
@@ -0,0 +1,357 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
```
Comment (bwplotka) [inline pkg/dedup/chunk_iter.go:3]: For other reviewers: Those things were moved, and (almost?) nothing was changed

Wonder if we can do the move in separate PR if we have time for easier review 🤔

File: test/e2e/compact_test.go [hunk 12]
```diff
@@ -537,6 +548,49 @@ func TestCompactWithStoreGateway(t *testing.T) {
 		{Value: 120, Metric: map[model.LabelName]model.LabelValue{"a": "1", "b": "5", "case": "full-replica-overlap-dedup-ready", "replica": "1"}},
 	}
 
+	if penaltyDedup {
```
Comment (bwplotka) [inline test/e2e/compact_test.go:551]: hah, nice trick

**Review by bwplotka — COMMENTED** [review]
I love it, I think the only thing that needs more work is:

* Small bits on flag description and documentation
* I did not reviewed `dedup/chunk_iter.go` as most of the code most moved (I seen it before ;p) Can you point us to what was changed what was not OR maybe do a move in separate PR? 🤔 

Thanks!

I think it's pretty solid and looks good to me, but let's check the iter code

File: pkg/dedup/chunk_iter.go (hunk 11 above)
Comment (yeya24) [reply pkg/dedup/chunk_iter.go:3]: Almost the same as compact merger. But it supports downsampled encoding. Please review from the overlapMerger struct. Thanks

File: cmd/thanos/compact.go (hunk 7 above)
Comment (yeya24) [reply cmd/thanos/compact.go:316]: Good idea!
But How can we detect unsupported labels? As it is experimental so users have to take the risk and backup data first?

File: pkg/dedup/chunk_iter.go (hunk 11 above)
Comment (GiedriusS) [reply pkg/dedup/chunk_iter.go:3]: I don't think GitHub supports it but locally you can do `git diff --color-moved main...`. It makes it clear what code has been simply moved :wink:

Comment (yeya24) [reply pkg/dedup/chunk_iter.go:3]: The core part is https://github.com/thanos-io/thanos/pull/4239/files#diff-b17d8397d84790d382e887eae8d7718260a36c1ff75d71150e9306437a3f90c9R153-R357. PTAL. They are used to vertically merge series.

File: pkg/compact/compact.go (hunk 10 above)
Comment (yeya24) [reply pkg/compact/compact.go:47]: Done.

File: cmd/thanos/compact.go (hunk 8 above)
Comment (yeya24) [reply cmd/thanos/compact.go:641]: Done.

File: cmd/thanos/compact.go (hunk 9 above)
Comment (yeya24) [reply cmd/thanos/compact.go:647]: Done.

File: cmd/thanos/compact.go (hunk 2 above)
Comment (kakkoyun) [reply cmd/thanos/compact.go:650]: It looks like it's still hidden, no?

File: cmd/thanos/compact.go (hunk 7 above)
Comment (kakkoyun) [reply cmd/thanos/compact.go:316]: @yeya24 I think Bartek talks about dedupFunc value.

**Review by kakkoyun — COMMENTED** [review]
I think this really close to the finish line.

File: cmd/thanos/compact.go (hunk 2 above)
Comment (yeya24) [reply cmd/thanos/compact.go:650]: @kakkoyun Do I need to do it in this pr?

File: cmd/thanos/compact.go (hunk 7 above)
Comment (yeya24) [reply cmd/thanos/compact.go:316]: I see. 😓

**Review by kakkoyun — APPROVED** [review]
Awesome! LGTM! :1st_place_medal: 

Let's give @bwplotka on last chance to review and we can merge.

**Comment by yeya24** [comment]
Friendly ping @bwplotka.

File: pkg/dedup/chunk_iter.go [hunk 13]
```diff
@@ -0,0 +1,357 @@
...
+	maxt := countChk.MaxTime
+
+	var (
+		chks [5]chunkenc.Chunk
+		chk  *chunks.Meta
+		err  error
+	)
+
+	chks[downsample.AggrCount] = countChk.Chunk
+	chk, err = a.toChunk(downsample.AggrSum, mint, maxt)
```
Comment (GiedriusS) [inline pkg/dedup/chunk_iter.go:269]: This block here probably needs a for-loop instead of copy/paste :smiley:

File: test/e2e/compact_test.go [hunk 14]
```diff
@@ -537,6 +548,49 @@ func TestCompactWithStoreGateway(t *testing.T) {
...
+			{Value: 120, Metric: map[model.LabelName]model.LabelValue{"a": "1", "b": "5", "case": "compaction-ready-one-block-marked-for-no-compact", "replica": "1"}},
+			{Value: 120, Metric: map[model.LabelName]model.LabelValue{"a": "1", "b": "6", "case": "compaction-ready-one-block-marked-for-no-compact", "replica": "1"}},
+			{Value: 120, Metric: map[model.LabelName]model.LabelValue{"a": "1", "b": "2", "case": "compaction-ready-after-dedup"}},
+			{Value: 120, Metric: map[model.LabelName]model.LabelValue{"a": "1", "b": "3", "case": "compaction-ready-after-dedup"}},
+			{Value: 120, Metric: map[model.LabelName]model.LabelValue{"a": "1", "b": "4", "case": "compaction-ready-after-dedup"}},
+			{Value: 120, Metric: map[model.LabelName]model.LabelValue{"a": "1", "b": "5", "case": "compaction-ready-after-dedup"}},
+			{Value: 120, Metric: map[model.LabelName]model.LabelValue{"a": "1", "b": "6", "case": "compaction-ready-after-dedup", "replica": "1"}},
+			{Value: 120, Metric: map[model.LabelName]model.LabelValue{"a": "1", "b": "1", "case": "a-partial-overlap-dedup-ready"}},
+			// If no penalty dedup enabled, the value should be 360.
```
Comment (GiedriusS) [inline test/e2e/compact_test.go:572]: Thanks for keeping these ordered, makes reviewing easier

File: cmd/thanos/compact.go [hunk 15]
```diff
@@ -623,15 +643,20 @@ func (cc *compactConfig) registerFlag(cmd extkingpin.FlagClause) {
 		Default("48h").SetValue(&cc.deleteDelay)
 
 	cmd.Flag("compact.enable-vertical-compaction", "Experimental. When set to true, compactor will allow overlaps and perform **irreversible** vertical compaction. See https://thanos.io/tip/components/compact.md/#vertical-compactions to read more."+
```
Comment (GiedriusS) [inline cmd/thanos/compact.go:645]: Let's add a space here: 
```suggestion
	cmd.Flag("compact.enable-vertical-compaction", "Experimental. When set to true, compactor will allow overlaps and perform **irreversible** vertical compaction. See https://thanos.io/tip/components/compact.md/#vertical-compactions to read more. "+
```

**Review by GiedriusS — COMMENTED** [review]
Just a few nits but overall LGTM :hugs:

File: docs/components/compact.md [hunk 16]
```diff
@@ -101,12 +101,12 @@ There can be few valid use cases for vertical compaction:
...
 This is very common while using [Receivers](../components/receive.md) with replication greater than 1 as receiver replication copies exactly the same timestamps and values to different receive instances.
-  * `realistic` duplication is when same series data is **logically duplicated**. For example, it comes from the same application, but scraped by two different Prometheus-es. Ideally
+  * `penalty` deduplication is when same series data is **logically duplicated**. For example, it comes from the same application, but scraped by two different Prometheus-es. Ideally
 this requires more complex deduplication algorithms. For example one that is used to [deduplicate on the fly on the Querier](query.md#run-time-deduplication-of-ha-groups). This is common
-case when Prometheus HA replicas are used. [Offline deduplication for this is in progress](https://github.com/thanos-io/thanos/issues/1014).
+case when Prometheus HA replicas are used. You can enable this deduplication via `--deduplication.func=penalty` flag.
```
[the remark this answers is not in the document]
Comment (yeya24) [reply docs/components/compact.md:109]: I am not sure about the 1% case. I only know it is related to counter reset/ rate functions. Is this https://github.com/thanos-io/thanos/issues/2890 the right issue for it? 

@bwplotka @kakkoyun for more input.

File: cmd/thanos/compact.go [hunk 17]
```diff
@@ -623,15 +643,20 @@ func (cc *compactConfig) registerFlag(cmd extkingpin.FlagClause) {
...
 		"NOTE: This flag is ignored and (enabled) when --deduplication.replica-label flag is set.").
 		Hidden().Default("false").BoolVar(&cc.enableVerticalCompaction)
 
+	cmd.Flag("deduplication.func", "Experimental. Deduplication algorithm for merging overlapping blocks. "+
+		"Possible values are: \"\", \"penalty\". If no value is specified, the default compact deduplication merger is used, which performs 1:1 deduplication for samples. "+
+		"When set to penalty, penalty based deduplication algorithm will be used. At least one replica label has to be set via --deduplication.replica-label flag.").
+		Default("").EnumVar(&cc.dedupFunc, compact.DedupAlgorithmPenalty, "")
+
 	cmd.Flag("deduplication.replica-label", "Label to treat as a replica indicator of blocks that can be deduplicated (repeated flag). This will merge multiple replica blocks into one. This process is irreversible."+
 		"Experimental. When it is set to true, compactor will ignore the given labels so that vertical compaction can merge the blocks."+
```
[the remark this answers is not in the document]
Comment (yeya24) [reply cmd/thanos/compact.go]: Done in https://github.com/thanos-io/thanos/pull/4239/commits/d0cba248a2914669389d31d06e517327df272b6d

File: pkg/dedup/chunk_iter.go (hunk 13 above)
Comment (yeya24) [reply pkg/dedup/chunk_iter.go]: Done in https://github.com/thanos-io/thanos/pull/4239/commits/d0cba248a2914669389d31d06e517327df272b6d

File: cmd/thanos/compact.go (hunk 15 above)
Comment (yeya24) [reply cmd/thanos/compact.go]: Done in https://github.com/thanos-io/thanos/pull/4239/commits/d0cba248a2914669389d31d06e517327df272b6d

File: docs/components/tools.md [hunk 18]
```diff
@@ -777,7 +777,7 @@ Flags:
       --rewrite.to-relabel-config-file=<file-path>
                                 Path to YAML file that contains relabel configs
                                 that will be applied to blocks
-      --tmp.dir="/tmp/thanos-rewrite"
+      --tmp.dir="/var/folders/0v/x6r6mgjx14ldyx7tfmpmx3gw0000gn/T/thanos-rewrite"
```
Comment (GiedriusS) [inline docs/components/tools.md]: Something happened here :scream:

Comment (yeya24) [reply docs/components/tools.md]: updated in a later commit

File: docs/components/compact.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (GiedriusS) [reply docs/components/compact.md:109]: Maybe let's say something like:

```
even though the deduplication algorithm has been battle-tested for quite a long time but still very few issues come up from time to time such as https://github.com/thanos-io/thanos/issues/2890
```

? It would be more concrete. And it wouldn't sound like we know that it is bad in some 1% of cases & we aren't fixing it :smile:

File: pkg/dedup/chunk_iter_test.go [hunk 19]
```diff
@@ -0,0 +1,325 @@
...
+			),
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			merged := m(tc.input...)
+			require.Equal(t, tc.expected.Labels(), merged.Labels())
+			actChks, actErr := storage.ExpandChunks(merged.Iterator())
+			expChks, expErr := storage.ExpandChunks(tc.expected.Iterator())
+
+			require.Equal(t, expErr, actErr)
```
Comment (GiedriusS) [inline pkg/dedup/chunk_iter_test.go:140]: BTW, why are you using this instead of `testutil.Equals`?

Comment (yeya24) [reply pkg/dedup/chunk_iter_test.go:140]: Good catch. These tests are copied from Prometheus and it uses `require`.

Comment (yeya24) [reply pkg/dedup/chunk_iter_test.go]: Done.

File: docs/components/compact.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (yeya24) [reply docs/components/compact.md:109]: Updated docs. PTAL

**Comment by yeya24** [comment]
I feel it is ready to go unless @bwplotka has some comments.
Will merge tomorrow if there is no objection.
